api/python/cellxgene_census/src/cellxgene_census/experimental/ml/cell_dataset_builder.py

```diff
@@ -0,0 +1,106 @@
+import uuid
+from abc import ABC, abstractmethod
+from typing import Any, Dict, Generator, Optional
+
+import scipy.sparse
+from datasets import Dataset
+from tiledbsoma import Experiment, ExperimentAxisQuery
+
+from cellxgene_census.experimental.util import X_sparse_iter
+
+
+class CellDatasetBuilder(ExperimentAxisQuery[Experiment], ABC):  # type: ignore
+    """
+    Abstract base class for methods to process CELLxGENE Census ExperimentAxisQuery
+    results into a Hugging Face Dataset in which each item represents one cell.
+    Subclasses implement the `cell_item()` method to process each row of an X layer
+    into a Dataset item, and may also override `__init__()` and context `__enter__()`
+    to perform any necessary preprocessing.
+
+    The base class inherits ExperimentAxisQuery, so typical usage would be:
+
+    ```
+    import cellxgene_census
+    import tiledbsoma
+    from cellxgene_census.experimental.ml import GeneformerTokenizer
+
+    with cellxgene_census.open_soma() as census:
+        with SubclassOfCellDatasetBuilder(
+            census["census_data"]["homo_sapiens"],
+            obs_query=tilebsoma.AxisQuery(...),  # define some subset of Census cells
+            ... # other ExperimentAxisQuery parameters e.g. var_query
+        ) as builder:
+            dataset = builder.build()
+    ```
+    """
+
+    def __init__(
+        self,
+        experiment: Experiment,
+        measurement_name: str = "RNA",
+        layer_name: str = "raw",
+        *,
+        _cells_per_chunk: int = 100_000,
+        **kwargs: Any,
+    ):
+        """
+        Initialize the CellDatasetBuilder to process the results of a Census
+        ExperimentAxisQuery.
+
+        - `experiment`: Census Experiment to be queried.
+        - `measurement_name`: Measurement in the experiment, default "RNA".
+        - `layer_name`: Name of the X layer to process, default "raw".
+        - `kwargs`: passed through to `ExperimentAxisQuery()`, especially `obs_query`
+           and `var_query`.
+        """
+        super().__init__(experiment, measurement_name, **kwargs)
+        self.layer_name = layer_name
+        self._cells_per_chunk = _cells_per_chunk
+
+    def build(self, from_generator_kwargs: Optional[Dict[str, Any]] = None) -> Dataset:
```

Comment:
To be tested in the near future @mlin `Dataset.from_generator` has a `num_proc` argument. We should test if when > 1 there's any conflicts with the thread pools from `X_sparse_iter`.

In my experience the multi-threading from `Dataset` methods work great and scale well with large data.

Reply:
Tagging @atolopko-czi to see if he has any thoughts as he may have (?) some working experience with this in PyTorch